File: Leetcode/1128. Number of Equivalent Domino Pairs/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn num_equiv_domino_pairs(dominoes: Vec<Vec<i32>>) -> i32 {
        let mut cache: HashMap<(i32, i32), i32> = HashMap::new();
        for domino in &dominoes {
            let (a, b) = (domino[0], domino[1]);
            let pair = if a < b { (a, b) } else { (b, a) };
            *cache.entry(pair).or_insert(0) += 1 ;
        }

        cache.values().map(|&value| value * (value - 1) / 2).sum()
    }
}

File: Leetcode/1128. Number of Equivalent Domino Pairs/solution.py
class Solution:
    def numEquivDominoPairs(self, dominoes: List[List[int]]) -> int:
        pairs = 0
        cache = dict()
        for a, b in dominoes:
            pair = (a, b) if a < b else (b, a)
            cache[pair] = cache.get(pair, 0) + 1

        return sum([value*(value-1)//2 for value in cache.values() if value > 0])